src/Microsoft.SymbolStore.Client/WindowsSymbolSever.cs

```diff
@@ -0,0 +1,234 @@
+﻿using System;
+using System.IO;
+using System.Net;
+using System.Threading.Tasks;
+
+namespace Microsoft.SymbolStore.Client
+{
+    public class SymbolServerResult
+    {
+        public Stream Stream { get; set; }
+        public long ContentLength { get; set; }
+        public bool Compressed { get; set; }
+
+        public SymbolServerResult(Stream result, long contentLength, bool compressed)
+        {
+            Stream = result;
+            ContentLength = contentLength;
+            Compressed = compressed;
+        }
+        public SymbolServerResult(WebResponse response, bool compressed)
+        {
+            Stream = response.GetResponseStream();
+            ContentLength = response.ContentLength;
+            Compressed = compressed;
+        }
+
+        public SymbolServerResult(string fileName)
+        {
+            Stream = File.OpenRead(fileName);
+            ContentLength = new FileInfo(fileName).Length;
+            Compressed = false;
+        }
+    }
+
+    public interface ISymbolServer
+    {
+        bool PreferThisServer { get; set; }
+
+        bool IsRemoteServer { get; }
+        bool IsReachable();
+        bool IsReachable(int timeout);
+        
+        SymbolServerResult FindPEFile(string filename, int buildTimeStamp, int imageSize);
+        Task<SymbolServerResult> FindPEFileAsync(string filename, int buildTimeStamp, int imageSize);
+        SymbolServerResult FindPdb(string pdbName, Guid guid, int age);
+        Task<SymbolServerResult> FindPdbAsync(string pdbName, Guid guid, int age);
+    }
+
+    public class WindowsSymbolSever : ISymbolServer
```

Comment:
I was doing some experiments with alternate implementations/refactorings of the SymbolServer here:
https://github.com/noahfalk/symstore/commit/0bbd743582e45b9798196d28bf85c25ea159b6a8

Its incomplete and some parts of it I'd certainly change, but here are the parts I think could be interesting:
a) SymbolCaches are a sub-type SymbolStore. All SymbolStores let you find files, SymbolCaches also let you add files.
b) I added the concept of a FileIdentity. You can give a key to any SymbolStore and it will give you back an identity. Symbol caches are allowed to assume that if two searches produce the same FileIdentity, then they will have the same content. In an ideal world the key itself would be a good identity, but in reality different servers give different content for the same key (ie symweb). I used the URI of the file on a particular server as the FileIdentity but that isn't required.
c) I created two implementations of the cache. One works like the current caches which ignores FileIdentity and allows you to store at most one item with the same key. The other can store content per-FileIdentity, even when there is more than one FileIdentity used for the same key. This lets you re-use the same cache for different symbol servers without any loss of fidelity about what came from where. Currently its not back-compatible with the old cache layout, but it could be made compatible by arbitrarily choosing one FileIdentity as the primary one for each key and storing it in the expected location. It also makes too many copies of files, but something like hard-links could probably remedy the needless duplicates.
d) Once you have a cache that can hold more than one file per-key, you can do file selection after caching rather than before. The UnionSymbolServer does it on a first response basis, but you could also rank sources in priority order or examine the content that came back to decide what is the best one to use.
e) I dealt almost entirely in terms of opaque keys. I added a few extension methods that offer convenience routines to consume the strongly typed file identifiers, but I would guess 95+% of the logic is file format agnostic. I assumed the keys were in a/b/c format, but even that doesn't seem like a hard requirement. I know you were predicting working with arbitrary keys was going to make it harder, but I'm not sure I follow your concerns on that one. From my experimenting so far I came to the opposite conclusion.